File: src/main/java/xyz/emirdev/echogen/commands/FlyCommand.java
package xyz.emirdev.echogen.commands;

import org.bukkit.entity.Player;

import com.mojang.brigadier.context.CommandContext;
import com.mojang.brigadier.exceptions.CommandSyntaxException;
import com.mojang.brigadier.tree.LiteralCommandNode;

import io.papermc.paper.command.brigadier.CommandSourceStack;
import io.papermc.paper.command.brigadier.Commands;
import io.papermc.paper.command.brigadier.argument.ArgumentTypes;
import net.kyori.adventure.text.minimessage.tag.resolver.Placeholder;
import net.kyori.adventure.text.minimessage.tag.resolver.TagResolver;
import xyz.emirdev.echogen.PluginCommand;

public class FlyCommand extends PluginCommand {

    @Override
    public LiteralCommandNode<CommandSourceStack> getCommand() {
        return Commands.literal("fly")
                .requires(ctx -> ctx.getSender().hasPermission("echogen.fly"))
                .executes(this::fly)
                .then(Commands.argument("player", ArgumentTypes.player())
                        .requires(ctx -> ctx.getSender().hasPermission("echogen.fly.others"))
                        .executes(this::flyOther))
                .build();
    }

    private int fly(CommandContext<CommandSourceStack> ctx) {
        Player player = getContextPlayer(ctx);
        if (player == null)
            return 1;

        boolean allowFlight = player.getAllowFlight();

        player.setAllowFlight(!allowFlight);

        if (!allowFlight)
            player.sendRichMessage("<aqua>You have <green>enabled</green> flight mode.</aqua>");
        else
            player.sendRichMessage("<aqua>You have <red>disabled</red> flight mode.</aqua>");

        return 1;
    }

    private int flyOther(CommandContext<CommandSourceStack> ctx) throws CommandSyntaxException {
        Player player = getContextPlayer(ctx);
        if (player == null)
            return 1;

        Player target = getPlayer(ctx, "player");

        boolean allowFlight = target.getAllowFlight();

        target.setAllowFlight(!allowFlight);

        TagResolver playerTag = Placeholder.unparsed("player", target.getName());

        if (!allowFlight) {
            player.sendRichMessage("<aqua>You have <green>enabled</green> flight mode for <player>.</aqua>", playerTag);
            target.sendRichMessage("<aqua>Your flight mode was <green>enabled</green>.</aqua>");
        } else {
            player.sendRichMessage("<aqua>You have <red>disabled</red> flight mode for <player>.</aqua>", playerTag);
            target.sendRichMessage("<aqua>Your flight mode was <red>disabled</red>.</aqua>");
        }

        return 1;
    }
}

File: src/main/kotlin/cat/emir/echogen/commands/FlyCommand.kt
package cat.emir.echogen.commands

import cat.emir.echogen.Echogen
import com.mojang.brigadier.context.CommandContext

import io.papermc.paper.command.brigadier.CommandSourceStack
import io.papermc.paper.command.brigadier.argument.ArgumentTypes
import cat.emir.echolib.command.PluginCommand
import cat.emir.echolib.command.getPlayer
import cat.emir.echolib.command.getPlayers
import cat.emir.echolib.sendLangMessage
import com.mojang.brigadier.builder.LiteralArgumentBuilder

class FlyCommand(plugin: Echogen) : PluginCommand<Echogen>(plugin) {

    override fun getCommand(): LiteralArgumentBuilder<CommandSourceStack> {
        return command("fly") {
            requires { it.sender.hasPermission("echogen.fly") }
            executes(::fly)
            argument("players", ArgumentTypes.players()) {
                requires { it.sender.hasPermission("echogen.fly.others") }
                executes(::flyOther)
            }
        }
    }

    fun fly(ctx: CommandContext<CommandSourceStack>): Int {
        val player = ctx.getPlayer() ?: return 1
        val allowFlight = player.allowFlight

        player.allowFlight = !allowFlight

        if (!allowFlight)
            player.sendLangMessage("fly.self.enabled")
        else
            player.sendLangMessage("fly.self.disabled")

        return 1
    }

    fun flyOther(ctx: CommandContext<CommandSourceStack>): Int {
        val sender = ctx.source.sender
        val targets = ctx.getPlayers("players") ?: return 1

        for (target in targets) {
            val allowFlight = target.allowFlight

            target.allowFlight = !allowFlight

            val data = listOf("player" to target.name)

            if (!allowFlight) {
                sender.sendLangMessage("fly.other.enabled.executor", data)
                target.sendLangMessage("fly.other.enabled.target")
            } else {
                sender.sendLangMessage("fly.other.disabled.executor", data)
                target.sendLangMessage("fly.other.disabled.target", data)
            }
        }

        return 1
    }
}
